gbm/windowing: fix compile error on some devices, cast to EGLNativeWi…

AreaScout opened the pull request:
## Description
cast to EGLNativeWindowType to fix compile error

``/home/odroid/hdd/xbmc/xbmc/windowing/gbm/WinSystemGbmGLESContext.cpp:122:83: error: no matching function for call to ‘CEGLContextUtils::CreatePlatformSurface(gbm_surface*, gbm_surface*)``

## How Has This Been Tested?

building on ODROID-N1, test using DRMPRIMEGLES render method -> works :)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Cosmetic change (non-breaking change that doesn't touch code)
- [ ] None of the above (please explain below)

## Checklist:
- [x] My code follows the **[Code Guidelines](CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

AreaScout committed:
gbm/windowing: fix compile error on some devices, cast to EGLNativeWindowType

pkerling commented:
What is your `EGLNativeWindowType`?

AreaScout commented:
https://github.com/rockchip-linux/libmali/blob/29mirror/include/FBDEV/eglplatform.h#L73
https://github.com/rockchip-linux/libmali/blob/29mirror/include/EGL/eglplatform.h#L74
https://github.com/mdrjr/5422_mali/blob/543eb787af5719224c2231490d5ebda71c1ec7a8/headers/EGL/eglplatform.h#L104

and khronos itself is using void* instead of gbm_surface:

https://www.khronos.org/registry/EGL/api/EGL/eglplatform.h

pkerling commented:
> It only compiles with the X11 library header files, the library itself is GBM enabled and x11 fbdev enabled

So broken vendor EGL libraries again, great... From a strict POV (and if I'm not mistaken), when `EGLNativeWindowType` is `Window` (an integer) we could silently lose data (i.e. get random startup crashes) by casting a pointer to it if the width of `Window` is smaller than the pointer width of the platform.

@lrusak What's your take on this? As it's gbm, I'll leave the decision up to you.

> and khronos itself is using void* instead of gbm_surface:

Yes that's fine, pointers implicitly convert to `void*`, so we don't have to cast. libmali is wrong here ofc defining it as `gbm_surface*`, but luckily it works anyway.

AreaScout commented:
at the end all are just pointers like void* so the cast should be fine and cover all type defines out there

Kwiboo commented:
For Rockchip LibreELEC we have `typedef struct gbm_surface * EGLNativeWindowType;` and that compiles and works without this cast.
https://github.com/rockchip-linux/libmali/blob/12daf22c405a4f8faf6cbc4d2e88b85b36dc61d9/include/EGL/eglplatform.h#L74 is the exact `eglplatform.h` used.

AreaScout commented:
yes that's true but 'struct gbm_surface * EGLNativeWindowType ' is also defined by the headers we use for X11 and i had problems with some X11 applications compiling, on XU4 we can't define gbm_surface to EGLNativeWindowType either it will not fail on kodi if we do but fail on others, that cast will not destroy anything as EGLNativeWindowType is always just a pointer so something like a struct or different and if i am not mistaken we use it already somewhere in the code

Kwiboo commented:
> So broken vendor EGL libraries again, great...

This is not broken EGL libraries, GBM was added just 4 months ago to Khronos `eglplatform.h` and "Khronos STRONGLY RECOMMENDS that you use the default definitions", vendors are free to define a `EGLNativeWindowType` suitable for the platform.

This cast will work for libmali platforms that have a `eglplatform.h` predating official definition.

AreaScout commented:
window is a pointer to an structure like gbm_surface is also a pointer to an structure the error is there because the compile thinks it has different structures underneath but it fact that doesn't matter because no one seems to use that EGLNativeWindowType type define everyone uses window in an x11 application or gbm_surface in gbm/drm application so the cast is ok and no one has to change headers here

pkerling commented:
> This is not broken EGL libraries, GBM was added just 4 months ago to Khronos eglplatform.h and "Khronos STRONGLY RECOMMENDS that you use the default definitions", vendors are free to define a EGLNativeWindowType suitable for the platform.

OK

> This cast will work for libmali platforms that have a eglplatform.h predating official definition.

I don't think this is the problem (do correct me if I'm wrong, I'm actually not that sure what's going on here) - we are handing a `gbm_surface*` to a function that wants `EGLNativeWindowType`, and these are the options we're discussing:

1. If `EGLNativeWindowType` is `void*`, code will just work since `gbm_surface*` is implicitly convertable to `void*`
1. If `EGLNativeWindowType` is `gbm_surface*`, code will just work since - well we're handing a `gbm_surface*` to a `gbm_surface*` parameter.

So... why do we need the cast except when `EGLNativeWindowType` is neither `void*` nor `gbm_surface*`? And if it is neither, would you think that is "suitable for the platform"?

pkerling commented:
> window is a pointer to an structure

No. `Window` is an integer that X11 uses as an identifier. See `X.h` on your system.

AreaScout commented:
but it want break X11 nor Kodi and if the headers from khronos are just 4 month old with this type define then we should use that cast for compatibility reasons

Kwiboo commented:
> So... why do we need the cast except when EGLNativeWindowType is neither void* nor gbm_surface*? And if it is neither, would you think that is "suitable for the platform"?

Correct the cast should not be needed when a `eglplatform.h` matching the compiled libmali is used.

From what I can tell libmali treats `EGLNativeWindowType` as a pointer for blobs that have been compiled with drm/dummy/fbdev winsys support, see 
https://github.com/rockchip-linux/libmali/tree/028bb21272bb3d2b81305b20bec2ed30bfdd2a6b/include/midgard/arm and https://github.com/rockchip-linux/libmali/tree/028bb21272bb3d2b81305b20bec2ed30bfdd2a6b/include/utgard/EGL for some internal include files of libmali.

AreaScout commented:
Ah ok, yes i agree on newest headers this is not needed, on kodi side everything is ok that i always say, it's just old headers, well i know now how to get around that, if you wont merge it, it's ok, but it wouldn't hurt

pkerling commented:
@lrusak If you want to merge, I highly recommend at least adding `static_assert(sizeof(EGLNativeWindowType) == sizeof(gbm_surface*))` so we know we can do the cast without messing up the data

AreaScout commented:
@pkerling 

added

Rechi commented on xbmc/windowing/gbm/WinSystemGbmGLESContext.cpp:
```diff
@@ -119,7 +119,9 @@ bool CWinSystemGbmGLESContext::CreateNewWindow(const std::string& name,
   }
 
-  if (!m_pGLContext.CreatePlatformSurface(m_GBM->GetSurface(), m_GBM->GetSurface()))
+  static_assert(sizeof(EGLNativeWindowType) == sizeof(gbm_surface*));
```
`static_assert` without a message isn't valid C++11, it was added in C++17.

a1rwulf commented:
Can we just close this one, or is anybody eager to merge?

AreaScout commented:
I tried to rebase this now, it's not working on my local branch seems I screwed it up, closing and reopen a new one

a1rwulf commented:
Anyway, let's wait for @lrusak or @Kwiboo to decide.

AreaScout commented:
I have tried to sync my local branch with current muster an force push it, but it won't rebase because of to much conflicts ... so I make a new one, I think this is important if someone still has "older" headers it wouldn't compile or in worst case leaking, thanks for notify me on this :)

notspiff commented:
uhm, nobody cares how you create the branch, whether you make it by hand or
use rebase. just force push it into the branch tracked by the PR. there is
absolutely no reason to make another PR (other than not understanding how
git/github works).

AreaScout commented:
that function is completely gone in master now and moved to a new different cpp file, this is what I got
https://pastebin.com/v9R8urMM

I couldn't resolve the merge conflict, rebase from an PR where the file has changed in master and moved to a new one is new to me, so yes in that case I do not know how git works.

I have checked out a new branch from header, can I merge the new changes in this PR somehow ?

AreaScout commented:
no i think it's to late I have already deleted my old branch ... :confused:

Rechi commented:
As you already have deleted the xbmc repo on GitHub for whatever reason, no.
Just deleting the branch wouldn't have been a problem.

The following changes would have fixed your PR without using `git rebase`
```
git checkout <YOUR-PR-BRANCH>
git reset --hard upstream/master
git add/git commit
git push -f origin <YOUR-PR-BRANCH>
```

notspiff commented:
a PR tracks a branch 'please pull from this branch'. you can stick whatever you feel like in that branch when you force push (you are rewriting the history). so no reason to open a new PR. nothing cares how you create the contents of the branch, rebase is just a convenience-mechanism as such.

AreaScout commented:
@Rechi thank you, I didn't know that 'git checkout <YOUR-PR-BRANCH>' is updating the PR branch from master, although it seems logical to me now.

@notspiff 
too late, I have deleted my fork already, next time I know

Rechi commented:
git checkout just checks out the branch (nothing else like you think now). I just mentioned it as I don't know which branch you have currently checked out.

AreaScout commented:
my PR branch is gone -> exterminated ! in short -> dead, I have changed my PR's a lot in the past and push forced it then, but keep local branch in sync with master when file is changed AND moved to new one which I do not have in my PR branch I don't know, sorry.

This is what I am using normally but to keep sync with master you need to rebase first and this can be tricky some times as it was in this case, but maybe you know it better.

this is what i use normally ``git commit -a --amend --no-edit``

notspiff commented:
@AreaScout that was the idea (you knowing next time.. :))
